# Introduction of Data Cleaning in Exploratory Data Analysis

## Initial Data Inspection

Before analysis or cleaning, it is useful to print a few rows of data. This helps ensure that the data is properly loaded. 

It also allows us to compare the observed data to the data dictionary and determine whether the coding appears to match our expectations. 

We can do so with .head() method

## Data Information

Once we’ve taken a first look at some data, a common next step is to address questions such as:

* How many (non-null) observations do we have?
* How many unique columns/features do we have?
* Which columns (if any) contain missing data?
* What is the data type of each column?

we can easily address these questions using the .info() method

## Inspecting Missing Data

After identifying that there is some missing data and converting it to a format that Python can recognize, it’s often a good idea to take a closer look at those rows. 

Sometimes, we can find clues as to WHY the data is missing, which can help us make decisions about whether to get rid of the rows altogether or impute the missing values somehow.

To perform such finding missing data:

syntax:

In [ ]:
df_name[df_name.isnull().any(axis=1)]

This will return every corresponding row data that contain missing value in any columns

## Example Study:

First, we should import the "pandas" module:

In [4]:
import pandas as pd
import numpy as np

Then, we need to import the .csv file to python using pandas:

In [5]:
df = pd.read_csv("diabetes.csv")

Below is the metadata of the dataset:
    
* Pregnancies: Number of times pregnant
* Glucose: Plasma glucose concentration at 2 hours in an oral glucose tolerance test
* BloodPressure: Diastolic blood pressure
* SkinThickness: Triceps skinfold thickness
* Insulin: 2-Hour serum insulin
* BMI: Body mass index
* DiabetesPedigreeFunction: Diabetes pedigree function
* Age: Age (years)
* Outcome: Class variable (0 or 1)

From the metadata, we can expect that 

* The datatype of Pregnancies is "int"
* The datatype of Glucose is "int/float"
* The datatype of BloodPressure is "float"
* The datatype of SkinThickness is "float"
* The datatype of Insulin is "int/float"
* The datatype of BMI is "int/float"
* The datatype of DiabetesPedigreeFunction is "int/float"
* The datatype of Age is "int"
* The datatype of Outcome is "bool"

### Inital Data Inspection

Let's take a look at the first 5 columns of the dataset

In [6]:
df.head()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1


Then we look at the information about each column: 
* datatypes
* number of observations
* number of columns
* any missing values

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


Lastly, we can take a look at the aggregate statistic of each data column:

In [8]:
df.describe(include="all")

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
count,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000
mean,3.845052,120.894531,69.105469,20.536458,79.799479,31.992578,0.471876,33.240885,0.348958
std,3.369578,31.972618,19.355807,15.952218,115.244002,7.884160,0.331329,11.760232,0.476951
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.078000,21.000000,0.000000
25%,1.000000,99.000000,62.000000,0.000000,0.000000,27.300000,0.243750,24.000000,0.000000
50%,3.000000,117.000000,72.000000,23.000000,30.500000,32.000000,0.372500,29.000000,0.000000
75%,6.000000,140.250000,80.000000,32.000000,127.250000,36.600000,0.626250,41.000000,1.000000
max,17.000000,199.000000,122.000000,99.000000,846.000000,67.100000,2.420000,81.000000,1.000000


The <include="all"> arugment specify to include all columns in the dataset.

From the Initial Data Inspection with 
* .head()
* .info
* .dtypes

We know that:

* There are 768 observations
* There are 9 columns
* The datatype of each columns
* Although there is no missing values in each columns, but we notices there are some 0 in some column that it should not be 0.
    * For example: Glucose, BloodPressure, SkinThickness, Insulin, BMI should not be 0
    * These values also seem to be way off from their respective medians and means, another indicator that something is off.
    * One way to interpret this is that these are missing values in the data.
* There are some columns that contains possible outliers.
    * For example: Insulin(max:846), Pregnancies(max:17), BMI (max:67.1)


### Inspecting Missing Data

Since there are some 0 in some values in Glucose, BloodPressure, SkinThickness, Insulin, BMI columns where it shouldn't, we decided to treat them as missing data.
Now we should replace 0s in those column with NaN to represent missing value

In [9]:
df[["Glucose", "BloodPressure", "SkinThickness","Insulin", "BMI"]] = df[["Glucose", "BloodPressure", "SkinThickness","Insulin", "BMI"]].replace(0, np.nan)

Now we should examine how many missing value are there in each column:

In [10]:
df.isna().sum()

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64

Now, we dive deep into those data rows with mising values. Let's print out the row data with missing values:

In [11]:
df[df.isna().any(axis=1)]

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148.0,72.0,35.0,NaN,33.6,0.627,50,1
1,1,85.0,66.0,29.0,NaN,26.6,0.351,31,0
2,8,183.0,64.0,NaN,NaN,23.3,0.672,32,1
5,5,116.0,74.0,NaN,NaN,25.6,0.201,30,0
7,10,115.0,NaN,NaN,NaN,35.3,0.134,29,0
...,...,...,...,...,...,...,...,...,...
761,9,170.0,74.0,31.0,NaN,44.0,0.403,43,1
762,9,89.0,62.0,NaN,NaN,22.5,0.142,33,0
764,2,122.0,70.0,27.0,NaN,36.8,0.340,27,0
766,1,126.0,60.0,NaN,NaN,30.1,0.349,47,1


From all the data missing rows:

* Majority of the data rows has more than 1 missing values
* The values in "Insulin" column are all NaN
    * This directly tells us those patient didnt have their Insulin recorded
    * further update is required for this dataset

Now that we investigated the missing values, we should start look at if the datatype of each data column has the correct datatype (numerical/categorical)

### Variable Type Analysis:

Let's take a look at the current datatype of each columns:

In [12]:
df.dtypes

Pregnancies                   int64
Glucose                     float64
BloodPressure               float64
SkinThickness               float64
Insulin                     float64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object

From the metadata and the data rows, we can expect that 

* The variable type of Pregnancies is "numeric"
* The variable type of Glucose is "numerical"
* The variable type of BloodPressure is "numeric"
* The variable type of SkinThickness is "numeric"
* The variable type of Insulin is "numeric"
* The variable type of BMI is "numeric"
* The variable type of DiabetesPedigreeFunction is "numeric"
* The variable type of Age is "numeric"
* The variable type of Outcome is "categorical(binary)"

We see that the "Outcome" column is represented as integer instead of categorical. WE should convert it from integer type to catgeorical type.

In [13]:
df["Outcome"] = pd.Categorical(
    df["Outcome"],
    [0,1],
    ordered=True
)

In [14]:
df.head()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148.0,72.0,35.0,NaN,33.6,0.627,50,1
1,1,85.0,66.0,29.0,NaN,26.6,0.351,31,0
2,8,183.0,64.0,NaN,NaN,23.3,0.672,32,1
3,1,89.0,66.0,23.0,94.0,28.1,0.167,21,0
4,0,137.0,40.0,35.0,168.0,43.1,2.288,33,1


In [15]:
df.dtypes

Pregnancies                    int64
Glucose                      float64
BloodPressure                float64
SkinThickness                float64
Insulin                      float64
BMI                          float64
DiabetesPedigreeFunction     float64
Age                            int64
Outcome                     category
dtype: object

Now the variable type of "Outcome" is cateogrical